src/proxy-ui-api/src/main/java/org/niis/xroad/restapi/service/ServiceClientService.java

```diff
@@ -0,0 +1,241 @@
+        EndpointType endpointType = endpointService.getServiceBaseEndpoint(serviceType);
+
+        List<AccessRightType> accessRightsByEndpoint = accessRightService
+                .getAccessRightsByEndpoint(clientType, endpointType);
+        return accessRightService.mapAccessRightsToServiceClients(clientType, accessRightsByEndpoint);
+    }
+
+    /**
+     * Get access right holders (serviceClients) for Endpoint
+     * @param id
+     * @return
+     * @throws EndpointNotFoundException    if no endpoint is found with given id
+     * @throws ClientNotFoundException      if client attached to endpoint is not found
+     */
+    public List<ServiceClientDto> getServiceClientsByEndpoint(Long id)
+            throws EndpointNotFoundException, ClientNotFoundException {
+
+        ClientType clientType = clientRepository.getClientByEndpointId(id);
+        EndpointType endpointType = endpointService.getEndpoint(id);
+
+        List<AccessRightType> accessRightsByEndpoint = accessRightService
+                .getAccessRightsByEndpoint(clientType, endpointType);
+        return accessRightService.mapAccessRightsToServiceClients(clientType, accessRightsByEndpoint);
+    }
+
+    /**
+     * Get service clients access rights to given client
+     *
+     * @param clientid
+     * @param serviceClientId
+     * @return
+     * @throws ClientNotFoundException if given client is not found
+     * @throws LocalGroupNotFoundException if given local group is not found
+     */
+    public List<ServiceClientAccessRightDto> getServiceClientAccessRights(ClientId clientid,
+            ServiceClientIdentifierDto serviceClientId) throws ClientNotFoundException, LocalGroupNotFoundException {
+        ClientType clientType = clientRepository.getClient(clientid);
+        if (clientType == null) {
+            throw new ClientNotFoundException("Client not found with id: " + clientid.toShortString());
+        }
+
+        // Get XRoadId for the given service client
+        XRoadId scId = serviceClientId.isLocalGroup()
+                ? localGroupService.getLocalGroupIdAsXroadId(serviceClientId.getLocalGroupId())
+                : serviceClientId.getXRoadId();
+
+        // Filter service clients access rights from the given clients acl-list
+        return clientType.getAcl().stream()
+                .filter(acl -> {
+                    boolean iseq = acl.getSubjectId().equals(scId);
+                    boolean isBaseEndpoint = acl.getEndpoint().isBaseEndpoint();
+                    return iseq && isBaseEndpoint;
+                })
+                .map(acl -> ServiceClientAccessRightDto.builder()
+                        .serviceCode(acl.getEndpoint().getServiceCode())
+                        .rightsGiven(FormatUtils.fromDateToOffsetDateTime(acl.getRightsGiven()))
+                        .title(getServiceTitle(clientType, acl.getEndpoint().getServiceCode()))
+                        .build())
+                .collect(Collectors.toList());
+    }
```

Comment:
Old implementation sorts by serviceCode, should this one do the same?

```
    services.sort! do |x, y|
      x[:service_code] <=> y[:service_code]
    end
```

Reply:
I guess it could but it's easy to do on frontend also, don't know if it is necessary to do here.

Reply:
It probably is not necessary to do that in backend. Frontend is probably just as good.